tracer/build/_build/Build.GitHub.cs

```diff
@@ -346,60 +412,96 @@ partial class Build
+                throw new Exception($"No builds found for {TargetBranch}");
+            }
+
+            BuildArtifact artifact = null;
+            var artifactName = $"{FullVersion}-release-artifacts";
+
+            Logger.Info($"Checking builds for artifact called: {artifactName}");
+
+            // start from the current commit, and keep looking backwards until we find a commit that has a build
+            // that has successful artifacts. Should only be called from branches with a linear history (i.e. single parent)
+            // This solves a potential issue where we previously selecting a build by start order, not by the actual
+            // git commit order. Generally that shouldn't be an issue, but if we manually trigger builds on master
+            // (which we sometimes do e.g. trying to bisect and issue, or retrying flaky test for coverage reasons),
+            // then we could end up selecting the wrong build.
+            const int maxCommitsBack = 20;
+            for (var i = 0; i < maxCommitsBack; i++)
+            {
+                var commitSha = GitTasks.Git($"log {TargetBranch}~{i} -1 --pretty=%H")
+                                        .FirstOrDefault(x => x.Type == OutputType.Std)
+                                        .Text;
+
+                Logger.Info($"Looking for builds for {commitSha}");
+
+                foreach (var build in builds)
+                {
+                    if (string.Equals(build.SourceVersion, commitSha, StringComparison.OrdinalIgnoreCase))
+                    {
+                        // Found a build for the commit, so should have an artifact
+                        try
+                        {
+                            artifact = await buildHttpClient.GetArtifactAsync(
+                                           project: AzureDevopsProjectId,
+                                           buildId: build.Id,
+                                           artifactName: artifactName);
+
+                            break;
+                        }
+                        catch (ArtifactNotFoundException)
+                        {
+                            Logger.Error($"Error: could not find {artifactName} artifact for build {build.Id} for commit {commitSha}. " +
+                                         $"Ensure the build has completed successfully for this commit before creating a release");
+                            throw;
+                        }
+                    }
+                }
+
+                if (artifact is not null)
+                {
+                    break;
+                }
+            }
+
+            if (artifact is null)
+            {
+                Logger.Error($"::error::Could not find artifacts called {artifactName} for release. Please manually trigger a fresh build of the pipeline if required.");
```

Comment:
> Please manually trigger a fresh build of the pipeline if required

Will that solve the issue? In the logic above, I thought we were only looking at builds triggered from CI, not manual builds

Reply:
Hmmmm, good point 🤔 Will need to work out a work around here. e.g.

- If the build has just _failed_ due to transient issues, then rerunning failed jobs should be fine, as counts as the same origin. But that doesn't help us if the CI failed to run at all for some reason (as we had recently with dodgy webhooks for example)
- Maybe we include manual runs too. If so, then we will have to _not_ fail (just skip) if artifacts aren't present for a manual run, as that could be normal.

Reply:
Chatting this through with @pierotibou we came up with the following:

- We should require successful builds for a release, just to be safe. Flaky builds will require a retry, but the number of these are reducing, we should be trying to tackle them, and they shouldn't be a massive blocker. 
- Allowing automatic releases from manual runs _generally_ could be a bit problematic, as you can manually change any parameters when creating the run, and not run tests. Probably won't be an issue, but maybe best to err on the safe side for now
- We should allow an escape hatch for releasing arbitrary commits/builds. We could add an optional CommitId/BuildId to the release process. If provided, the release will try and use these artifacts specifically. I think we would have to do _some_ validation of this to avoid mistakes, e.g. "is the commit on the current branch" but it will probably come in useful at some point. 

I'd like to do that latter point in a separate PR, as doesn't seem critical to do yet. Does that all sound reasonable to you @zacharycmontoya?